File: packages/ui-flutter/lib/src/data_grid.dart
import 'package:flutter/material.dart';

import 'app_text.dart';
import 'theme.dart';

/// A single DataGrid column: [id] identifies it for edit callbacks,
/// [cellText] both renders the cell and (when [sortable]) drives the sort
/// compare and seeds the edit field.
class KinetixDataGridColumn<T> {
  const KinetixDataGridColumn({
    required this.id,
    required this.header,
    required this.cellText,
    this.width = 120,
    this.sortable = false,
    this.editable = false,
    this.onCellEdit,
  });

  final String id;
  final String header;
  final double width;
  final bool sortable;
  final bool editable;
  final String Function(T row) cellText;
  final void Function(T row, int rowIndex, String value)? onCellEdit;
}

/// Mirrors `packages/ui/src/components/data-grid.tsx`'s row-virtualized
/// grid. Column resize, drag-to-reorder, and column pin are web-only: no
/// touch-friendly drag-a-column-border gesture convention on mobile, and a
/// sticky column needs a custom layout `ListView` doesn't give for free.
/// Ships the three features that map directly onto Flutter primitives
/// instead: `ListView.builder` only builds rows near the viewport (bound
/// its height via a parent `SizedBox`/`Expanded`, same as
/// `KinetixVirtualList`), a tap-to-sort header, and tap-to-edit cells that
/// commit on the keyboard's submit action rather than the web's
/// blur-also-commits. The header and body scroll horizontally in lockstep
/// via two `ScrollController`s kept in sync — a `ScrollController` can only
/// attach to one scrollable at a time in Flutter, so they can't just share
/// one.
class KinetixDataGrid<T> extends StatefulWidget {
  const KinetixDataGrid({
    super.key,
    required this.columns,
    required this.rows,
    this.rowHeight = 40,
  });

  final List<KinetixDataGridColumn<T>> columns;
  final List<T> rows;
  final double rowHeight;

  @override
  State<KinetixDataGrid<T>> createState() => _KinetixDataGridState<T>();
}

class _KinetixDataGridState<T> extends State<KinetixDataGrid<T>> {
  final ScrollController _headerController = ScrollController();
  final ScrollController _bodyController = ScrollController();
  final TextEditingController _editController = TextEditingController();
  String? _sortColumnId;
  bool _ascending = true;
  int? _editingRow;
  String? _editingColumnId;

  @override
  void initState() {
    super.initState();
    _bodyController.addListener(() {
      if (_headerController.hasClients && _headerController.offset != _bodyController.offset) {
        _headerController.jumpTo(_bodyController.offset);
      }
    });
  }

  @override
  void dispose() {
    _headerController.dispose();
    _bodyController.dispose();
    _editController.dispose();
    super.dispose();
  }

  List<T> get _sortedRows {
    final columnId = _sortColumnId;
    if (columnId == null) return widget.rows;
    final column = widget.columns.firstWhere((c) => c.id == columnId);
    final sorted = [...widget.rows]..sort((a, b) => column.cellText(a).compareTo(column.cellText(b)));
    return _ascending ? sorted : sorted.reversed.toList();
  }

  void _toggleSort(KinetixDataGridColumn<T> column) {
    if (!column.sortable) return;
    setState(() {
      if (_sortColumnId == column.id) {
        _ascending = !_ascending;
      } else {
        _sortColumnId = column.id;
        _ascending = true;
      }
    });
  }

  double get _totalWidth => widget.columns.fold(0, (sum, c) => sum + c.width);

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    final sorted = _sortedRows;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        SingleChildScrollView(
          controller: _headerController,
          scrollDirection: Axis.horizontal,
          physics: const NeverScrollableScrollPhysics(),
          child: SizedBox(
            width: _totalWidth,
            height: widget.rowHeight,
            child: DecoratedBox(
              decoration: BoxDecoration(
                color: c.background,
                border: Border(bottom: BorderSide(color: c.border)),
              ),
              child: Row(
                children: [
                  for (final column in widget.columns)
                    GestureDetector(
                      behavior: HitTestBehavior.opaque,
                      onTap: () => _toggleSort(column),
                      child: SizedBox(
                        width: column.width,
                        child: Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 8),
                          child: Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Flexible(
                                child: Text(
                                  column.header,
                                  overflow: TextOverflow.ellipsis,
                                  style: AppText.labelLg.copyWith(color: c.mutedForeground),
                                ),
                              ),
                              if (_sortColumnId == column.id) ...[
                                const SizedBox(width: 4),
                                Icon(
                                  _ascending ? Icons.arrow_upward : Icons.arrow_downward,
                                  size: 12,
                                  color: c.mutedForeground,
                                ),
                              ],
                            ],
                          ),
                        ),
                      ),
                    ),
                ],
              ),
            ),
          ),
        ),
        Expanded(
          child: ListView.builder(
            controller: _bodyController,
            itemCount: sorted.length,
            itemExtent: widget.rowHeight,
            itemBuilder: (context, rowIndex) {
              final row = sorted[rowIndex];
              return DecoratedBox(
                decoration: BoxDecoration(border: Border(bottom: BorderSide(color: c.border))),
                child: SizedBox(
                  width: _totalWidth,
                  child: Row(
                    children: [
                      for (final column in widget.columns) _cell(context, column, row, rowIndex),
                    ],
                  ),
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _cell(BuildContext context, KinetixDataGridColumn<T> column, T row, int rowIndex) {
    final c = KinetixTheme.of(context);
    final isEditing = _editingRow == rowIndex && _editingColumnId == column.id;
    return SizedBox(
      width: column.width,
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 8),
        child: isEditing
            ? TextField(
                controller: _editController,
                style: AppText.bodyMd.copyWith(color: c.foreground),
                decoration: const InputDecoration(border: InputBorder.none, isDense: true),
                onSubmitted: (value) {
                  column.onCellEdit?.call(row, rowIndex, value);
                  setState(() {
                    _editingRow = null;
                    _editingColumnId = null;
                  });
                },
              )
            : GestureDetector(
                behavior: HitTestBehavior.opaque,
                onTap: column.editable
                    ? () {
                        setState(() {
                          _editingRow = rowIndex;
                          _editingColumnId = column.id;
                          _editController.text = column.cellText(row);
                        });
                      }
                    : null,
                child: Text(
                  column.cellText(row),
                  overflow: TextOverflow.ellipsis,
                  style: AppText.bodyMd.copyWith(color: c.foreground),
                ),
              ),
      ),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/DataGrid.swift
//
// DataGrid.swift — KinetixDataGrid.
//
// Mirrors packages/ui/src/components/data-grid.tsx's row-virtualized grid.
// Column resize, drag-to-reorder, and column pin are web-only: no
// touch-friendly drag-a-column-border gesture convention on iOS, and a
// sticky column needs a custom layout `ScrollView` doesn't give for free.
// Ships the three features that map directly onto SwiftUI primitives
// instead: a `LazyVStack` inside a `ScrollView` only instantiates rows near
// the viewport (bind the viewport via `.frame(height:)` at the call site,
// same as KinetixVirtualList), a `Section` header pinned via
// `pinnedViews: [.sectionHeaders]` stays fixed at the top while the whole
// `ScrollView` pans in both directions — free horizontal sync between
// header and body, no manual scroll-offset bridging needed — and editable
// cells swap to a `TextField` on tap, committing on Return (`.onSubmit`)
// rather than the web's blur-also-commits.
//

import SwiftUI

public struct KinetixDataGridColumn<Row> {
    public let id: String
    public let header: String
    public let width: CGFloat
    public let sortable: Bool
    public let editable: Bool
    public let cellText: (Row) -> String
    public let onCellEdit: ((Row, Int, String) -> Void)?

    public init(
        id: String,
        header: String,
        width: CGFloat = 120,
        sortable: Bool = false,
        editable: Bool = false,
        cellText: @escaping (Row) -> String,
        onCellEdit: ((Row, Int, String) -> Void)? = nil
    ) {
        self.id = id
        self.header = header
        self.width = width
        self.sortable = sortable
        self.editable = editable
        self.cellText = cellText
        self.onCellEdit = onCellEdit
    }
}

public struct KinetixDataGrid<Row>: View {
    @Environment(\.kinetixColors) private var colors

    private let columns: [KinetixDataGridColumn<Row>]
    private let rows: [Row]
    private let rowHeight: CGFloat

    @State private var sortColumnId: String?
    @State private var sortAscending = true
    @State private var editing: (rowIndex: Int, columnId: String)?
    @State private var draft: String = ""

    public init(columns: [KinetixDataGridColumn<Row>], rows: [Row], rowHeight: CGFloat = 40) {
        self.columns = columns
        self.rows = rows
        self.rowHeight = rowHeight
    }

    private var sortedRows: [Row] {
        guard
            let sortColumnId,
            let column = columns.first(where: { $0.id == sortColumnId })
        else { return rows }
        let ascending = rows.sorted { column.cellText($0) < column.cellText($1) }
        return sortAscending ? ascending : ascending.reversed()
    }

    private func toggleSort(_ column: KinetixDataGridColumn<Row>) {
        guard column.sortable else { return }
        if sortColumnId == column.id {
            sortAscending.toggle()
        } else {
            sortColumnId = column.id
            sortAscending = true
        }
    }

    private var headerRow: some View {
        HStack(spacing: 0) {
            ForEach(columns, id: \.id) { column in
                Button {
                    toggleSort(column)
                } label: {
                    HStack(spacing: 4) {
                        Text(column.header)
                            .font(.kinetixLabelLg)
                        if sortColumnId == column.id {
                            Image(systemName: sortAscending ? "chevron.up" : "chevron.down")
                                .font(.system(size: 10))
                        }
                    }
                    .foregroundStyle(colors.mutedForeground)
                    .frame(width: column.width, height: rowHeight, alignment: .leading)
                    .padding(.horizontal, 8)
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .disabled(!column.sortable)
            }
        }
        .background(colors.background)
        .overlay(alignment: .bottom) {
            Rectangle().fill(colors.border).frame(height: 1)
        }
    }

    @ViewBuilder
    private func cell(column: KinetixDataGridColumn<Row>, row: Row, rowIndex: Int) -> some View {
        let isEditing = editing?.rowIndex == rowIndex && editing?.columnId == column.id
        Group {
            if isEditing {
                TextField("", text: $draft)
                    .font(.kinetixBody)
                    .textFieldStyle(.plain)
                    .onSubmit {
                        column.onCellEdit?(row, rowIndex, draft)
                        editing = nil
                    }
            } else {
                Text(column.cellText(row))
                    .font(.kinetixBody)
                    .foregroundStyle(colors.foreground)
                    .lineLimit(1)
                    .onTapGesture {
                        guard column.editable else { return }
                        draft = column.cellText(row)
                        editing = (rowIndex, column.id)
                    }
            }
        }
        .frame(width: column.width, alignment: .leading)
        .padding(.horizontal, 8)
    }

    public var body: some View {
        ScrollView([.horizontal, .vertical]) {
            LazyVStack(alignment: .leading, spacing: 0, pinnedViews: [.sectionHeaders]) {
                Section(header: headerRow) {
                    ForEach(Array(sortedRows.enumerated()), id: \.offset) { rowIndex, row in
                        HStack(spacing: 0) {
                            ForEach(columns, id: \.id) { column in
                                cell(column: column, row: row, rowIndex: rowIndex)
                            }
                        }
                        .frame(height: rowHeight)
                        .overlay(alignment: .bottom) {
                            Rectangle().fill(colors.border).frame(height: 1)
                        }
                    }
                }
            }
        }
        .border(colors.border)
    }
}
